flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/sink/committer/FilesCommitter.java

```diff
@@ -0,0 +1,314 @@
+      } else {
+        commitDeltaTxn(committableCollection, dataFilesNum, deleteFilesNum);
+      }
+    }
+  }
+
+  private void replacePartitions(
+      Collection<CommitRequest<FilesCommittable>> writeResults,
+      int dataFilesNum,
+      int deleteFilesNum) {
+    // Partition overwrite does not support delete files.
+    Preconditions.checkState(deleteFilesNum == 0, "Cannot overwrite partitions with delete files.");
+
+    // Commit the overwrite transaction.
+    ReplacePartitions dynamicOverwrite = table.newReplacePartitions().scanManifestsWith(workerPool);
+
+    String flinkJobId = null;
+    long checkpointId = -1;
+    for (CommitRequest<FilesCommittable> commitRequest : writeResults) {
+      WriteResult result = commitRequest.getCommittable().committable();
+      Preconditions.checkState(
+          result.referencedDataFiles().length == 0, "Should have no referenced data files.");
+      flinkJobId = commitRequest.getCommittable().jobID();
+      checkpointId = commitRequest.getCommittable().checkpointId();
+      Arrays.stream(result.dataFiles()).forEach(dynamicOverwrite::addFile);
+    }
+
+    commitOperation(
+        writeResults,
+        dynamicOverwrite,
+        dataFilesNum,
+        0,
+        "dynamic partition overwrite",
+        flinkJobId,
+        checkpointId);
+  }
+
+  private void commitDeltaTxn(
+      Collection<CommitRequest<FilesCommittable>> writeResults,
+      int dataFilesNum,
+      int deleteFilesNum) {
+    if (deleteFilesNum == 0) {
+      // To be compatible with iceberg format V1.
+      AppendFiles appendFiles = table.newAppend().scanManifestsWith(workerPool);
+      String flinkJobId = null;
+      long checkpointId = -1;
+      for (CommitRequest<FilesCommittable> commitRequest : writeResults) {
+        WriteResult result = commitRequest.getCommittable().committable();
+        Preconditions.checkState(
+            result.referencedDataFiles().length == 0, "Should have no referenced data files.");
+        Arrays.stream(result.dataFiles()).forEach(appendFiles::appendFile);
+        flinkJobId = commitRequest.getCommittable().jobID();
+        checkpointId = commitRequest.getCommittable().checkpointId();
+      }
+
+      commitOperation(
+          writeResults, appendFiles, dataFilesNum, 0, "append", flinkJobId, checkpointId);
+
+    } else {
+      // To be compatible with iceberg format V2.
+      for (CommitRequest<FilesCommittable> commitRequest : writeResults) {
+        WriteResult result = commitRequest.getCommittable().committable();
+        RowDelta rowDelta = table.newRowDelta().scanManifestsWith(workerPool);
+        Arrays.stream(result.dataFiles()).forEach(rowDelta::addRows);
+        Arrays.stream(result.deleteFiles()).forEach(rowDelta::addDeletes);
+
+        String flinkJobId = commitRequest.getCommittable().jobID();
+        long checkpointId = commitRequest.getCommittable().checkpointId();
+        commitOperation(
```

Comment:
I think it is intentional to commit the changes one-by-one, so we know which delete file applies to which data file, but it would be good to leave a comment here to describe why we do this, so the developers reading this code later will not "fix" this issue - even better if we have a test for this as well.

Reply:
Added some comment & UT